lib/wheel.c

```diff
@@ -149,6 +155,73 @@ int wheel_add_item(struct timer_wheel *wheel, void *item)
 			   slot % wheel->slots);
 	listnode_add(wheel->wheel_slot_lists[slot % wheel->slots], item);
 
+	/*
+	 * Figure out how many slots we should skip with the newly added
+	 * item
+	 */
+	while (list_isempty(
+		       wheel->wheel_slot_lists[(curr_slot + slots_to_skip) %
+					       wheel->slots]) &&
+	       (curr_slot + slots_to_skip) % wheel->slots != curr_slot)
+		slots_to_skip++;
+
+	if (debug_timer_wheel)
+		zlog_debug("Old slots to skip: %u, new slots to skip: %" PRIu64,
+			   wheel->slots_to_skip, slots_to_skip);
+
+	/*
+	 * The wheel timer has refigured out that the number of slots
+	 * to skip hasn't changed from the insertion.  There is
+	 * nothing to change here at all.  So let's just return
+	 */
+	if (wheel->slots_to_skip == slots_to_skip)
+		return 0;
+
+	/*
+	 * At this point the wheel timer knows that the insertion of the
+	 * new item is inbetween the old time to pop and the soon
+	 * to be new time to pop.  The wheel timer must figure out
+	 * when to pop given how much time has passed
+	 */
+	timer_remain = thread_timer_remain_msec(wheel->timer);
```

Comment:
asking 'timer_remain' is pretty expensive (it takes a mutex, e.g., for god knows what reason), and the computation around "time_reduction" and all that is a little complicated. so maybe we could just be conservative, assume we've waited almost long enough, and always just pop if we have an insertion? the 'thread' will be scheduled, so repeated insertions shouldn't mess that up? the wheel doesn't try to pace itself particularly carefully anyway - it doesn't try to account for drift between timer wakeups, it doesn't adjust for time spent doing work when visiting a slot, e.g.